Source/Dafny/Dafny.atg

```diff
@@ -1871,25 +1875,43 @@ Formals<.bool incoming, bool allowGhostKeyword, bool allowNewKeyword, List<Forma
      Type ty;
      bool isGhost;
      bool isOld;
+     Expression defaultValue;
   .)
   "("
   [
     GIdentType<allowGhostKeyword, allowNewKeyword, out id, out ty, out isGhost, out isOld>
-                 (. formals.Add(new Formal(id, id.val, ty, incoming, isGhost, isOld)); .)
+    ParameterDefaultValue<incoming, out defaultValue>
+                 (. formals.Add(new Formal(id, id.val, ty, incoming, isGhost, defaultValue, isOld)); .)
     { "," GIdentType<allowGhostKeyword, allowNewKeyword, out id, out ty, out isGhost, out isOld>
-                 (. formals.Add(new Formal(id, id.val, ty, incoming, isGhost, isOld)); .)
+    ParameterDefaultValue<incoming, out defaultValue>
+                 (. formals.Add(new Formal(id, id.val, ty, incoming, isGhost, defaultValue, isOld)); .)
     }
   ]
   ")"
   .
 
+ParameterDefaultValue<bool incoming, out Expression defaultValue>
+= (. defaultValue = null; IToken tok;
+  .)
+  [ ":="        (. tok = t; .)
+    Expression<out defaultValue, true, true>
+    (. if (!incoming) {
+         SemErr(tok, "out-parameters cannot have default-value expressions");
```

Comment:
For the sake of argument...couldn't they? :) The same syntax could be useful for initializing a default return value. Could be nice to add in the future.

Reply:
That's fair - I could see it being convenient for implementors, but they can just initialize the out parameters to their default values explicitly at the top of the method body instead, and that doesn't pollute the signature.